### DISCUSSION TITLE
Configuring minimum input length / user-defined callback whether results should show

### DISCUSSION THREAD

**Discussion by richardebeling** [orchidjs/tom-select#376] [Ideas]
I have a multi-select of names, and to prevent biases with my users, I want to only show matching results after at least 3 characters have been typed by the user.

### What doesn't work
* The [documentation](https://tom-select.js.org/docs/) lists the `shouldLoad` callback as a potential option:
    > Use the shouldLoad() callback to implement minimum input length or other input validation. 
    
   However, this only limits whether `load` will be called or not here: https://github.com/orchidjs/tom-select/blob/50f25e65ed6e0bd104d5bffb764cd1e15640e924/src/tom-select.ts#L749-L751
    any options that are already loaded (in my case: `option`s of the original `select`, so they are always loaded) will still be shown to users if `shouldLoad` returns `false`. ([jsbin](https://jsbin.com/badotukahi/1/edit?html,js,output))

* The documentation also lists the `score` callback as a potential option:
    > Overrides the scoring function used to sort available options. The provided function should return a function that returns a number greater than or equal to zero to represent the score of an item (the function's first argument). If 0, the option is declared not a match.
    
    However, tom-select internally uses sifter to perform the searching/scoring, and sifter has a hardcoded path for 0-length inputs that assigns a score of 1 to each element [here](https://github.com/brianreavis/sifter.js/blob/97270b46707ddfb117570653f74153919dbb9cd7/sifter.js#L364-L375). Thus, I can only limit the display _after_ the user has entered something ([jsbin](https://jsbin.com/qowemewaxi/1/edit?html,js,output)). For the initial dropdown, this might be solvable using `openOnFocus`, but this doesn't help if the user enters some filter and then removes it again. Also, the user can still manually open the dropdown with an empty filter.

### What could potentially work

* Sifter provides a [`limit` option](https://github.com/brianreavis/sifter.js#api) that truncates results to this length. If I were able to pass in this option through tomselect, e.g. by tomselect offering a `additionalSifterOptions` setting, this could be used to limit results display in user-defined ways. I think it is suboptimal, as this is only applied _after_ results have been scored, filtered and sorted, all of which wastes computing power if we already know that we do not want to show any results, but I guess it would work for me.

* Tomselect could offer some kind of callback setting that can be used to signal that no results should be shown. This could by implemented by adding a simple check [here](https://github.com/orchidjs/tom-select/blob/50f25e65ed6e0bd104d5bffb764cd1e15640e924/src/tom-select.ts#L1311):
    ```javascript
    if (self.settings.shouldShowResults && self.settings.shouldShowResults.call(self,query))) {
        return emptyResult; // build an emtpy sifter result here, or define a constant emtpyResult somewhere to be used here
    }
    ```

### Conclusion
Am I missing something, is there some other way to achieve this? If not, is there a chance any of the options above could be made available to users?

**Comment by oyejorge** [comment]
Interesting. Yeah, shouldLoad would be the obvious choice but it only effects remote data. We might need to change that.

[Here's another approach](https://jsbin.com/janijomoke/1/edit?html,js,output)

**Reply by Toxmarz** [reply]
This example does what I am trying to do, but I can't figure out how to incorporate the javascript code in the example with my Tom Select inputs. Any insight would be appreciated. Nothing I try does anything except maybe break them.

**Reply by Toxmarz** [reply]
Ok, nevermind. Been playing with it, I needed to run the example script BEFORE  the new TomSelect, I had it executing after (and even in the middle ?!? ) Thanks!

**Reply by Toxmarz** [reply]
Sorry, now that I have it working, the plugins no longer work. I was using the remove, clear and dropdown headers. Is it possilble to configure so these to work with this additional javascript function?

**Reply by Toxmarz** [reply]
Sorry, again nevermind. Figured it out. I thought the example code was added to an existing TomSelect input, but realized it replaces it completely and when got rid of the old instance I moved the plugins to this, works like a charm. As you might guess this is a little more advanced javascript for me.

 (function() {
  class MinimumLengthSelect extends TomSelect{    
    refreshOptions(triggerDropdown=true){
      var query	= this.inputValue();
      if( query.length < 4){
        this.close(false);
        return;
      }      
      super.refreshOptions(triggerDropdown);
    }    
  }
  
  new MinimumLengthSelect("#idOfYourSelect",{
    plugins: {
		  'remove_button':{
			  'title':'Remove this item',
		  },
			'clear_button':{
			  'title':'Remove all selected options',
		  },
			'dropdown_header': {
			  'title': 'Search/select zip codes...'
			}
    }
  });
})();

if ithis helps speeed things along for someone else
